### DISCUSSION TITLE
[Code generation] How do you add a trailing comma to a generated enum?

### DISCUSSION THREAD

**Discussion by dulcineachau** [dotnet/roslyn#73814] [Q&A] [answered]
I'm writing code that generates an enum value. But I can't figure out how to get the enum value to include a trailing comma. Is this supported? if so - how do we do it?

I'll give an example - let's say we are trying to generate a C# class called TestEnum.

I want:
```
public class TestEnum
{
     None = 0,
     One = 1,
     Two = 2,
}

```
What I get by default is:
```
public class TestEnum
{
     None = 0,
     One = 1,
     Two = 2
}
```

Imagine the code to create the enum class is something like this below where I manually created a member list:
```
            return SyntaxFactory.EnumDeclaration(
                attributeLists: SyntaxFactory.SingletonList<AttributeListSyntax>(null),
                modifiers: SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)),
                identifier: SyntaxFactory.Identifier(SystemTaskEnumName),
                baseList: null,
                members: memberList)
                .WithLeadingTrivia(header);
```

I already tried making it so memberList uses a SeparatedSyntaxList with an explicit list of separators matching the same size of the list. But that doesn't work (presumably because the signature requests for that list to be -1 size of the set). I've also tried searching around for any indications of trailing comma support on the net and haven't found any. So is this something that simply doesn't work / isn't supproted today?

**Comment by CyrusNajmabadi** [comment]
members is `SeparatedSyntaxList<EnumMemberDeclarationSyntax>`  These can be made with a SyntaxNodeOrTokenList, which itself just needs the individual members and commas in their proper order.

**Reply by dulcineachau** [reply]
I've already tried this - it doesn't seem to work. I construct this manually but the passed in separators list doesn't change the outputted results. 

Here's an example:
`
            memberList = SyntaxFactory.SeparatedList<EnumMemberDeclarationSyntax>(
                constructedMemberList,
                Enumerable.Repeat(SyntaxFactory.Token(SyntaxKind.CommaToken), constructedMemberList.Count()));
`

I even tried using a different delimiter (like a semicolon) just to validate whether it is honored - but any separator list passed in doesn't seem to actually be honored. It always uses a comma and truncates the last one.

I will note that the documentation for that constructor explicitly states:
        /// <param name="separators">A sequence of token to be interleaved between the nodes. The number of tokens must
        /// be one less than the number of nodes.</param>
        
Which seems to imply you can't override the trailing delimit on the last node.

**Reply by CyrusNajmabadi** [reply]
You need to make a SyntaxNodeOrTokenList and pass that into this function:

```c#
public static SeparatedSyntaxList<TNode> SeparatedList<TNode>(SyntaxNodeOrTokenList nodesAndTokens) where TNode : SyntaxNode
```

The nodes and tokens need to be interspersed.  So: Node, Token, Node, Token, etc.

**Reply by CyrusNajmabadi** [reply, marked as answer]
Here's an example of how we do it in roslyn: https://github.com/dotnet/roslyn/blob/e0abd64c4fcb7d574cfed766df66759a49bdf78f/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/CodeGeneration/EnumMemberGenerator.cs#L24-L51

When we add a member to an enum, we preserve a trailing comma if that's what the user already had.

**Reply by dulcineachau** [reply]
Ah got it. So I did try what you mentioned. Oddly this doesn't work for me even though it works for the Roslyn code. 

You can see in the below screen shot - I interspersed the member + comma as you mentioned. Processing of that is all fine and I verified every member has a comma added after it.
![image](https://github.com/dotnet/roslyn/assets/166661977/e0806e0f-0cdd-4165-a88c-641d2e86714a)

But as soon as I pass in that list to create a Separated List - the last separator gets dropped.
`
return SyntaxFactory.SeparatedList<EnumMemberDeclarationSyntax>(
     constructedMemberList);
`

The returned separated list's # of nodes vs # of separators does not match. The separators is one less then the List<SyntaxNodeOrToken> I originally constructed and passed in (original size 1660, some we get a Separated list of 830 + 829)
![image](https://github.com/dotnet/roslyn/assets/166661977/fe28b2f0-f2e9-4c00-818e-94000c0f051b)

**Reply by dulcineachau** [reply]
Ah I will note we are using version 3.11.0. So if this is a later feature - that might be why. I need to see if I can upgrade to a newer version and will try again.

**Reply by dulcineachau** [reply]
Nevermind! The debugging issue was a bug in my code. Using your mechanism worked. Thank you! Marked as answer.

**Reply by CyrusNajmabadi** [reply]
I'd have to see your code.
